src/charm.py

```diff
@@ -6,20 +6,355 @@
+        try:
+            json.loads(config)
+        except json.JSONDecodeError as e:
+            raise InvalidConfigError(f"Failed to decode json: {e}")
+
+    @property
+    def _config(self) -> _RouteConfig:
+        return _RouteConfig(rule=self.config.get("rule"), root_url=self.config.get("root_url"))
+
+    @property
+    def rule(self) -> Optional[str]:
+        """The Traefik rule this charm is responsible for configuring."""
+        return self._config.rule
+
+    @property
+    def root_url(self) -> Optional[str]:
+        """The advertised url for the charm requesting ingress."""
+        return self._config.root_url
+
+    def _render_config(self, model_name: str, unit_name: str, app_name: str) -> RouteConfig:
+        return self._config.render(model_name=model_name, unit_name=unit_name, app_name=app_name)
+
+    @staticmethod
+    def _generate_traefik_unit_config(config: RouteConfig) -> UnitConfig:
+        rule, config_id, url = config.rule, config.id_, config.root_url
+
+        traefik_router_name = f"juju-{config_id}-router"
+        traefik_service_name = f"juju-{config_id}-service"
+
+        config: "UnitConfig" = {
+            "router": {
+                "rule": rule,
+                "service": traefik_service_name,
+                "entryPoints": ["web"],
+                "middlewares": ["oathkeeper"],
+            },
+            "router_name": traefik_router_name,
+            "service": {"loadBalancer": {"servers": [{"url": url}]}},
+            "service_name": traefik_service_name,
+        }
+        return config
+
+    @staticmethod
+    def _merge_traefik_configs(configs: Iterable["UnitConfig"]) -> "TraefikConfig":
+        traefik_config = {
+            "http": {
+                "routers": {config["router_name"]: config["router"] for config in configs},
+                "services": {config["service_name"]: config["service"] for config in configs},
+            }
+        }
+        return traefik_config
+
+    def _get_relation(self, endpoint: str) -> Optional[Relation]:
+        """Fetches the Relation for endpoint and checks that there's only 1."""
+        relations = self.model.relations.get(endpoint)
+        if not relations:
+            logger.info(f"no relations yet for {endpoint}")
+            return None
+        if len(relations) > 1:
+            logger.warning(f"more than one relation for {endpoint}")
+        return relations[0]
+
+    @property
+    def _ipu_relation(self) -> Optional[Relation]:
+        """The relation with the unit requesting ingress."""
+        return self._get_relation(self._ingress_relation_name)
+
+    def _config_for_unit(self, unit_data: RequirerData) -> RouteConfig:
+        """Get the _RouteConfig for the provided `unit_data`."""
+        unit_name = unit_data["name"]
+        model_name = unit_data["model"]
+
+        # sanity checks
+        assert unit_name is not None, "remote unit did not provide its name"
+        assert "/" in unit_name, unit_name
```

Comment:
Just have a question about the assertion: Where do we disable it when we deploy the charm? Or we don't disable it?